### PR TITLE
ARROW-7842: [Rust] [Parquet] implement array_reader for list type columns

### PR DESCRIPTION
**Pull request by mcassels**
Nested lists are not yet supported. PR for new test file for datafusion test: apache/parquet-testing#11

### CODE REVIEW

**Comment by mcassels** [comment]
New test file PR for datafusion test: https://github.com/apache/parquet-testing/pull/11

**Comment by mcassels** [comment]
@andygrove @nevi-me do you have any thoughts on this?

File: rust/datafusion/src/logicalplan.rs [hunk 1]
```diff
@@ -828,8 +828,8 @@ mod tests {
         .build()?;
 
         let expected = "Projection: #id\
-        \n  Selection: #state Eq Utf8(\"CO\")\
-        \n    TableScan: employee.csv projection=Some([0, 3])";
+                        \n  Selection: #state Eq Utf8(\"CO\")\
+                        \n    TableScan: employee.csv projection=Some([0, 3])";
```
Comment (andygrove) [inline rust/datafusion/src/logicalplan.rs:833]: Are these formatting changes from running rustfmt?

File: rust/datafusion/src/utils.rs [hunk 2]
```diff
@@ -74,6 +74,29 @@ macro_rules! make_string {
     }};
 }
 
+macro_rules! make_string_from_list {
+    ($column: ident, $row: ident) => {{
+        let list = $column
+            .as_any()
+            .downcast_ref::<array::ListArray>()
+            .unwrap()
```
Comment (andygrove) [inline rust/datafusion/src/utils.rs:82]: `expect` would be better than `unwrap`. Using a `Result` would be better than `expect`.

File: rust/datafusion/src/utils.rs [hunk 3]
```diff
@@ -74,6 +74,29 @@ macro_rules! make_string {
...
+            .value($row);
+        let string_values = match (0..list.len())
+            .map(|i| array_value_to_string(list.clone(), i))
+            .collect::<Result<Vec<String>>>()
+        {
+            Ok(values) => values,
+            _ => {
+                return Err(ExecutionError::ExecutionError(format!(
+                    "Unsupported {:?} type for repl.",
```
Comment (andygrove) [inline rust/datafusion/src/utils.rs:91]: What does `repl` mean here? Maybe this error could be a bit more descriptive?

File: rust/datafusion/src/utils.rs [hunk 4]
```diff
@@ -120,6 +143,7 @@ pub fn array_value_to_string(column: array::ArrayRef, row: usize) -> Result<Stri
         DataType::Time64(unit) if *unit == TimeUnit::Nanosecond => {
             make_string!(array::Time64NanosecondArray, column, row)
         }
+        DataType::List(_) => make_string_from_list!(column, row),
```
Comment (andygrove) [inline rust/datafusion/src/utils.rs:146]: Does it matter what data type the list is? Can lists of any type be converted into strings?

**Review by andygrove — COMMENTED** [review]
Sorry for the late review. I wasn't aware of this PR until you mentioned me. My email filters didn't match the title ... this looks great overall. I left a few comments and would like to see some feedbackfrom someone more familiar with the parquet crate. I've requested some other reviewers take a look. Thanks for the contribution!

File: rust/datafusion/src/utils.rs (hunk 4 above)
Comment (mcassels) [reply rust/datafusion/src/utils.rs:139]: This recursively calls `array_value_to_string` on the array items, so an unsupported type should give the appropriate error on the recursive call

File: rust/datafusion/src/logicalplan.rs (hunk 1 above)
Comment (mcassels) [reply rust/datafusion/src/logicalplan.rs:833]: yes

File: rust/parquet/src/arrow/array_reader.rs [hunk 5]
```diff
@@ -468,6 +491,391 @@ where
...
+        ArrowType::Time32(ArrowTimeUnit::Millisecond) => {
+            build_empty_list_array_with_primitive_items!(ArrowTime32MillisecondType)
+        }
+        ArrowType::Time64(ArrowTimeUnit::Microsecond) => {
+            build_empty_list_array_with_primitive_items!(ArrowTime64MicrosecondType)
+        }
+        ArrowType::Time64(ArrowTimeUnit::Nanosecond) => {
+            build_empty_list_array_with_primitive_items!(ArrowTime64NanosecondType)
+        }
+        ArrowType::Duration(ArrowTimeUnit::Second) => {
```
Comment (nevi-me) [inline rust/parquet/src/arrow/array_reader.rs:587]: Are durations supported in parquet?

**Review by nevi-me — APPROVED** [review]
LGTM

File: rust/parquet/src/arrow/array_reader.rs (hunk 5 above)
Comment (mcassels) [reply rust/parquet/src/arrow/array_reader.rs:528]: It looks like the PrimitiveArrayReader implementation [supports Duration] (https://github.com/apache/arrow/blob/master/rust/parquet/src/arrow/array_reader.rs#L206)

Comment (nevi-me) [reply rust/parquet/src/arrow/array_reader.rs:528]: That's probably incorrect, but we can clarify/fix the behaviour of the readers at a later stage.

Comment (nevi-me) [reply rust/parquet/src/arrow/array_reader.rs:528]: That's probably incorrect, but we can clarify/fix the behaviour of the readers at a later stage.

**Comment by maxburke** [comment]
Is there anything else that needs to be taken care of in order to have this merged?

**Comment by nevi-me** [comment]
> Is there anything else that needs to be taken care of in order to have this merged?

We need to wait for https://github.com/apache/parquet-testing/pull/11 to be merged, then update the submodule so that the tests pass.

@sunchao @paddyhoran may you please also have a look at this when you get a chance. I have some slight concerns around Arrow <-> Parquet format support, but I'm comfortable that we'll be able to pick up any required changes when we progress with the Arrow to Parquet writer.

**Comment by nevi-me** [comment]
@mcassels may you please kindly rebase and update the submodule?

**Comment by maxburke** [comment]
@nevi-me Is this okay to merge now?

**Comment by maxburke** [comment]
+ @sunchao and @paddyhoran

**Comment by sunchao** [comment]
Sorry for the late response @maxburke . I'll take a look at this today.

**Comment by maxburke** [comment]
> Sorry for the late response @maxburke . I'll take a look at this today.

I hate to keep poking but were you able to have a look at this? We're really hoping to get back onto the master branch of Arrow

**Comment by maxburke** [comment]
(FWIW it seems that this build is broken because of the merge of ARROW-3827

**Comment by sunchao** [comment]
@maxburke yes was looking at this but the change was bigger than I expected and I was not able to finish it. Still looking.

**Comment by maxburke** [comment]
@paddyhoran have you been able to take a look at this?

**Comment by paddyhoran** [comment]
Sorry @maxburke, I'm snowed under at work at the moment.  Once @sunchao or @sadikovi approves we are good, they are the Parquet experts.

File: rust/parquet/src/arrow/array_reader.rs [hunk 6]
```diff
@@ -745,16 +1152,66 @@ impl<'a> TypeVisitor<Option<Box<dyn ArrayReader>>, &'a ArrayReaderBuilderContext
     }
 
     /// Build array reader for list type.
-    /// Currently this is not supported.
     fn visit_list_with_item(
```
Comment (sunchao) [inline rust/parquet/src/arrow/array_reader.rs:1155]: Is this being called anywhere?

File: rust/parquet/src/arrow/array_reader.rs [hunk 7]
```diff
@@ -426,6 +450,391 @@ where
...
+macro_rules! build_empty_list_array_with_non_primitive_items {
+    ($builder:ident) => {{
+        let values_builder = $builder::new(0);
+        let mut builder = ListBuilder::new(values_builder);
+        let empty_list_array = builder.finish();
+        Ok(Arc::new(empty_list_array))
+    }};
+}
+
+fn build_empty_list_array(item_type: ArrowType) -> Result<ArrayRef> {
```
Comment (sunchao) [inline rust/parquet/src/arrow/array_reader.rs:503]: The way this is implemented isn't ideal as it has to do pattern matching on every possible type. Is there a way to simplify this? For instance, every `ArrowPrimitiveType` already has a `DataType` as its field.

File: rust/parquet/src/arrow/array_reader.rs [hunk 8]
```diff
@@ -426,6 +450,391 @@ where
...
+        self
+    }
+
+    /// Returns data type.
+    /// This must be a List.
+    fn get_data_type(&self) -> &ArrowType {
+        &self.data_type
+    }
+
+    fn next_batch(&mut self, batch_size: usize) -> Result<ArrayRef> {
```
Comment (sunchao) [inline rust/parquet/src/arrow/array_reader.rs:746]: Does this handle nested lists?

File: rust/datafusion/src/logicalplan.rs [hunk 9]
```diff
@@ -1004,8 +1004,8 @@ mod tests {
         .build()?;
 
         let expected = "Projection: #state, #total_salary\
-        \n  Aggregate: groupBy=[[#state]], aggr=[[SUM(#salary) AS total_salary]]\
-        \n    TableScan: employee.csv projection=Some([3, 4])";
+             \n  Aggregate: groupBy=[[#state]], aggr=[[SUM(#salary) AS total_salary]]\
```
Comment (sunchao) [inline rust/datafusion/src/logicalplan.rs:1007]: Is this change related? same below

File: rust/datafusion/src/execution/physical_plan/expressions.rs [hunk 10]
```diff
@@ -39,9 +39,9 @@ use arrow::compute;
 use arrow::compute::kernels::arithmetic::{add, divide, multiply, subtract};
 use arrow::compute::kernels::boolean::{and, or};
 use arrow::compute::kernels::cast::cast;
-use arrow::compute::kernels::comparison::{eq, gt, gt_eq, lt, lt_eq, neq};
 use arrow::compute::kernels::comparison::{
-    eq_utf8, gt_eq_utf8, gt_utf8, like_utf8, lt_eq_utf8, lt_utf8, neq_utf8, nlike_utf8,
```
Comment (sunchao) [inline rust/datafusion/src/execution/physical_plan/expressions.rs:44]: Is this change related?

**Review by sunchao — COMMENTED** [review]
Sorry again for the late review. Have being out of touch for some time and is taking time to get back to the latest state of Arrow reader in parquet-rs. Left some comments but will probably add more later.

File: rust/parquet/src/arrow/array_reader.rs (hunk 6 above)
Comment (mcassels) [reply rust/parquet/src/arrow/array_reader.rs:1155]: Yeah, it's called here: https://github.com/apache/arrow/blob/master/rust/parquet/src/schema/visitor.rs#L51

File: rust/parquet/src/arrow/array_reader.rs (hunk 8 above)
Comment (mcassels) [reply rust/parquet/src/arrow/array_reader.rs:746]: No, this PR doesn't handle nested lists. At the end of `visit_list_with_item` there's a check on the item type and it gives an error if the List item type is List https://github.com/apache/arrow/pull/6770/files#diff-fca053527b552a9f2abb1e280e570cb6R1197

File: rust/datafusion/src/logicalplan.rs (hunk 9 above)
Comment (mcassels) [reply rust/datafusion/src/logicalplan.rs:1007]: no, this was from running cargo fmt

File: rust/datafusion/src/execution/physical_plan/expressions.rs (hunk 10 above)
Comment (mcassels) [reply rust/datafusion/src/execution/physical_plan/expressions.rs:44]: No, I can remove this change if you prefer. Just saw that there were 2 groups of imports from the same namespace and combined them to one group.

File: rust/parquet/src/arrow/array_reader.rs (hunk 7 above)
Comment (mcassels) [reply rust/parquet/src/arrow/array_reader.rs:503]: I don't know whether there's a way to do the association in the opposite direction. In this case we want to go from a DataType to an ArrowPrimitiveType. I found that pattern matching is currently used in other similar situations like https://github.com/apache/arrow/blob/master/rust/arrow/src/json/reader.rs#L433 for example

File: rust/parquet/src/arrow/array_reader.rs [hunk 11]
```diff
@@ -745,16 +1152,66 @@ impl<'a> TypeVisitor<Option<Box<dyn ArrayReader>>, &'a ArrayReaderBuilderContext
...
-        _context: &'a ArrayReaderBuilderContext,
+        context: &'a ArrayReaderBuilderContext,
     ) -> Result<Option<Box<dyn ArrayReader>>> {
-        Err(ArrowError(
-            "Reading parquet list array into arrow is not supported yet!".to_string(),
-        ))
+        let mut new_context = context.clone();
+
+        let list_child = &list_type.get_fields()[0];
+        let item_child = &list_child.get_fields()[0];
```
Comment (sunchao) [inline rust/parquet/src/arrow/array_reader.rs:1164]: This doesn't look correct. I think we should use the `_item_type` passed in for item type. List type in Parquet can be represented with several different kinds of schema as mentioned [here](https://github.com/apache/parquet-format/blob/master/LogicalTypes.md#lists). [This method](https://github.com/apache/arrow/blob/master/rust/parquet/src/schema/visitor.rs#L35) already helped to setup the list item type for you.

File: rust/parquet/src/arrow/array_reader.rs (hunk 8 above)
Comment (sunchao) [reply rust/parquet/src/arrow/array_reader.rs:746]: Ok. It will be good if we have a path to get there. Can also reference the C++ impl. Also could you add some comments in this to indicate that nested lists are not supported?

File: rust/parquet/src/arrow/array_reader.rs [hunk 12]
```diff
@@ -426,6 +450,391 @@ where
...
+            }
+        }
+        let batch_values = match null_list_indices.len() {
+            0 => next_batch_array.clone(),
+            _ => remove_indices(next_batch_array.clone(), item_type, null_list_indices)?,
+        };
+
+        // null list has def_level = 0
+        // empty list has def_level = 1
+        // null item in a list has def_level = 2
```
Comment (sunchao) [inline rust/parquet/src/arrow/array_reader.rs:779]: I think we also need to check if the list item is required or optional right? the list itself can also be either optional or requried.

File: rust/parquet/src/arrow/array_reader.rs [hunk 13]
```diff
@@ -426,6 +450,391 @@ where
...
+        // null lists have def_level = 0
+        let mut null_list_indices: Vec<usize> = Vec::new();
+        for i in 0..def_levels.len() {
+            if def_levels[i] == 0 {
+                null_list_indices.push(i);
+            }
+        }
+        let batch_values = match null_list_indices.len() {
+            0 => next_batch_array.clone(),
+            _ => remove_indices(next_batch_array.clone(), item_type, null_list_indices)?,
```
Comment (sunchao) [inline rust/parquet/src/arrow/array_reader.rs:774]: Hmm why you need to process the batch array and add them through a builder? why cannot just directly use the array in the result `ListArray` and set up validity bits and offsets only?

**Review by sunchao — CHANGES_REQUESTED** [review]
@mcassels thanks for your patience and I added more comments on this PR.

File: rust/arrow/src/ipc/convert.rs [hunk 14]
```diff
@@ -518,11 +518,12 @@ pub(crate) fn get_fb_field_type<'a: 'b, 'b>(
             )
         }
         List(ref list_type) => {
+            let field_name = fbb.create_string("list"); // field schema requires name to be not None
```
[the remark this answers is not in the document]
Comment (mcassels) [reply rust/arrow/src/ipc/convert.rs:521]: As I understand it this field name is not actually going to be used. Here, List fields are being treated like a Struct containing a single field whose children are each of the items in the list. This dummy "list" name is being used as the name for the single child field, not as the name for the List field itself.

[the remark this answers is not in the document]
Comment (mcassels) [reply rust/arrow/src/ipc/convert.rs:521]: The reason why I named it "list": "The outer-most level must be a group annotated with LIST that contains a single field named list" https://github.com/apache/parquet-format/blob/master/LogicalTypes.md#lists

File: rust/parquet/src/arrow/array_reader.rs (hunk 13 above)
Comment (mcassels) [reply rust/parquet/src/arrow/array_reader.rs:774]: As I understand it, the null values that represent a null **list** rather than a null list item need to not be present in the child data of the final ListArray. I don't like that this uses the builder either, but I wasn't sure how else to create that child data array that has specific indices excluded... I might be misunderstanding something though. Do you have any suggestions?

File: rust/parquet/src/arrow/array_reader.rs (hunk 11 above)
Comment (mcassels) [reply rust/parquet/src/arrow/array_reader.rs]: Yes, that makes more sense. I'm not sure why I did it that way before

File: rust/parquet/src/arrow/array_reader.rs (hunk 8 above)
Comment (mcassels) [reply rust/parquet/src/arrow/array_reader.rs:747]: I've just added a couple comments -- on the ArrayReader implementation and on visit_list_with_item

File: rust/parquet/src/arrow/array_reader.rs (hunk 12 above)
Comment (mcassels) [reply rust/parquet/src/arrow/array_reader.rs:786]: Yeah it seems that this comment is not correct... I think it would be correct just in the case where both the list and the list item are optional. 

However, much of the logic in this comment is not actually used: Rep and def levels are compared only with 0. Do you see any flaws in the def and rep level logic that comes after this comment? If the actual logic is ok I'll just remove the comment.

File: rust/parquet/src/arrow/array_reader.rs (hunk 11 above)
Comment (mcassels) [reply rust/parquet/src/arrow/array_reader.rs]: The failing test made me realize why I had it that way. Currently item_type is passed as `&Type` to `visit_list_with_item` while list_type is passed as `Rc<Type>`. Because item_type is not passed as `Rc<Type>`, if `item_type` is used then when the list items are visited by PrimitiveArrayReader, they will not satisfy the `self.is_included` check, because that checks using the reference. 

I've now changed the signature of `visit_list_with_item` to pass item_type as `Rc<Type>` to fix this issue.

File: rust/arrow/src/ipc/convert.rs (hunk 14 above)
[the remark this answers is not in the document]
Comment (sunchao) [reply rust/arrow/src/ipc/convert.rs:521]: That is Parquet specification though. Here you are creating a flat buffer object which should follow Arrow's specification. See [here](https://arrow.apache.org/docs/format/Columnar.html#schema-message).

File: rust/parquet/src/arrow/array_reader.rs (hunk 13 above)
Comment (sunchao) [reply rust/parquet/src/arrow/array_reader.rs:781]: Not sure if I understand you correctly. I think a null list is represented with the combination of valid bit array and offset array, and should have no presence in the value array (example [here](https://arrow.apache.org/docs/format/Columnar.html#variable-size-list-layout)). Therefore, the input array can just be passed intact to the ListArray as value array.

It would also be useful to refer the C++ [impl. ](https://github.com/apache/arrow/blob/master/cpp/src/parquet/arrow/reader_internal.cc#L1348)

**Comment by emkornfield** [comment]
@nevi-me did your other PRs supercede this?

**Comment by nevi-me** [comment]
I'll pick this up in the coming 2 weeks

**Comment by nevi-me** [comment]
This has sadly fallen behind significantly from the main branch. I've spent a few hours trying to rebase, but because it looks like the PR contained changes for the below, it's a bit difficult to follow. 

* `filter` kernel for lists
* `contains` kernel (also can't tell if it was introduced then removed in subsequent commits, but looks like that)
* the parquet list reader

I got as far as getting everything except the filter kernel working, but I suspect I'll be spending a few more hours dealing with merge conflicts; so I'm throwing in the towel for now. I'll try a different strategy of reimplementing the parquet changes on the main branch separately.

@yordan-pavlov would you have capacity to help with filter kernels for lists? We would use @mcassels work, but convert it to use the `FilterContext` that you introduced (IIRC).

**Comment by nevi-me** [comment]
> @nevi-me did your other PRs supercede this?

I hadn't seen this @emkornfield. No, I've only been focusing on the writer, so I'll still want to merge this into the branch currently called master.
I've also noticed that we could split the PR into 2 or 3, so I'll try do that as I've spent a few hours trying to rebase; and I'll have to throw away that time :(

**Comment by yordan-pavlov** [comment]
@nevi-me I am happy to help but I couldn't see any changes to the filter kernel in this PR; have I misunderstood your message?

**Comment by nevi-me** [comment]
> @nevi-me I am happy to help but I couldn't see any changes to the filter kernel in this PR; have I misunderstood your message?

Apologies, this PR came from a fork that UrbanLogiq maintain, so I think they had worked on list filters for their purposes, and then initially included those commits in this PR, but eventually reversed them.

I think it's still good to incorporate those changes, as the relevant JIRA has been open for a while (ARROW-5350).

I've extracted the changes and pushed them to my fork at https://github.com/nevi-me/arrow/commit/4c11f46a62e770ff3c56d3898daf429099a8cb1a. You'll see that I commented out the filter on lists, as I couldn't figure out how to make them work with `FilterContext` as mentioned.

There's already some tests, so looks like a solid base to start from. You can copy the changes from my commit, or push into my ARROW-5350 branch; either's fine with me.

**Comment by yordan-pavlov** [comment]
thanks @nevi-me , I understand now, will have a look this week

**Comment by mcassels** [comment]
> > @nevi-me I am happy to help but I couldn't see any changes to the filter kernel in this PR; have I misunderstood your message?
> 
> Apologies, this PR came from a fork that UrbanLogiq maintain, so I think they had worked on list filters for their purposes, and then initially included those commits in this PR, but eventually reversed them.
> 
> I think it's still good to incorporate those changes, as the relevant JIRA has been open for a while ([ARROW-5350](https://issues.apache.org/jira/browse/ARROW-5350)).
> 
> I've extracted the changes and pushed them to my fork at [nevi-me@4c11f46](https://github.com/nevi-me/arrow/commit/4c11f46a62e770ff3c56d3898daf429099a8cb1a). You'll see that I commented out the filter on lists, as I couldn't figure out how to make them work with `FilterContext` as mentioned.
> 
> There's already some tests, so looks like a solid base to start from. You can copy the changes from my commit, or push into my [ARROW-5350](https://issues.apache.org/jira/browse/ARROW-5350) branch; either's fine with me.

Yes we had worked on list filters as well but intended to make that a separate PR. But as you can see unfortunately we got busy and fell behind main :/

**Comment by velvia** [comment]
@nevi-me @yordan-pavlov so turns out this has become important to fix and merge for us too and we need to bring our branch up to date.  Let us know best way to go forward, but one of us is happy to start porting this change to a newer updated branch and let everyone have a look, and push a formal Arrow PR as well.

**Comment by nevi-me** [comment]
> @nevi-me @yordan-pavlov Hi so I joined UrbanLogiq recently and am helping to maintain this stuff. Turns out this has become important to fix and merge for us too and we need to bring our branch up to date. Let us know best way to go forward, but one of us is happy to start porting this change to a newer updated branch and let everyone have a look, and push a formal Arrow PR as well.

I've already ported the changes to a separate branch, and I'm currently going through @sunchao's review. I ran out of time over the weekend, but I might be able to complete the changes in the coming week.

I know it creates a dependency on me, but given the hours I've already spent on this; I'd rather still continue and address the review, as I might be able to merge the new PR immediately if I don't have questions for anyone.

**Comment by velvia** [comment]
> I know it creates a dependency on me, but given the hours I've already spent on this; I'd rather still continue and address the review, as I might be able to merge the new PR immediately if I don't have questions for anyone.

@nevi-me that sounds great, looking forward to it!  There is another feature we might look to push from our branch sometime.  :)

File: rust/arrow/src/ipc/convert.rs (hunk 14 above)
[the remark this answers is not in the document]
Comment (nevi-me) [reply rust/arrow/src/ipc/convert.rs:521]: This will be resolved separately as part of the integration work that I'm doing. I agree with Chao that this is unrelated to Parquet.

**Comment by nevi-me** [comment]
Closed in favour of #8449
